p2p/src/main/kotlin/maru/p2p/PeerDiscoveryTask.kt

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Consensys Software Inc.
+ *
+ * This file is dual-licensed under either the MIT license or Apache License 2.0.
+ * See the LICENSE-MIT and LICENSE-APACHE files in the repository root for details.
+ *
+ * SPDX-License-Identifier: MIT OR Apache-2.0
+ */
+package maru.p2p
+
+import java.time.Duration
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import maru.p2p.discovery.MaruDiscoveryPeer
+import maru.p2p.discovery.MaruDiscoveryService
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import org.apache.tuweni.bytes.Bytes
+import tech.pegasys.teku.networking.p2p.libp2p.PeerAlreadyConnectedException
+import tech.pegasys.teku.networking.p2p.network.P2PNetwork
+import tech.pegasys.teku.networking.p2p.peer.Peer
+import tech.pegasys.teku.networking.p2p.reputation.ReputationManager
+
+class PeerDiscoveryTask(
+  private val discoveryService: MaruDiscoveryService,
+  private val p2pNetwork: P2PNetwork<Peer>,
+  private val reputationManager: ReputationManager,
+  private val maxPeers: Int,
+  private val getPeerCount: () -> Int,
+) {
+  private val log: Logger = LogManager.getLogger(this.javaClass)
+
+  private val currentlySearching = AtomicBoolean(false)
+  private var discoveryTaskFuture: ScheduledFuture<*>? = null
+  private val started = AtomicBoolean(false)
+  private val connectionInProgress = mutableListOf<Bytes>()
+  private val scheduler = Executors.newSingleThreadScheduledExecutor(Thread.ofPlatform().daemon().factory())
```

Comment:
MInor: I suggest move it to constructor with the default. I consider instantiation of executors inside components as a technical debt. At some point in time, IMO, we should pass on the same pool to all the components